src/nunavut/lang/cpp/support/VariableLengthArray.hpp

```diff
@@ -0,0 +1,464 @@
+        {
+            // we ignore the exception since all allocation failures are modeled using
+            // null by this class.
+            (void) e;
+        }
+
+        if (new_data != nullptr)
+        {
+            if (new_data != data_)
+            {
+                move_and_free(new_data, data_, size_, capacity_, alloc_);
+            }  // else the allocator was able to simply extend the reserved area for the same memory pointer.
+            data_     = new_data;
+            capacity_ = no_shrink_capacity;
+        }
+
+        return capacity_;
+    }
+
+    /**
+     * Deallocate or reallocate memory such that not more than {@code size()} elements can be stored in this object.
+     *
+     * @return True if nothing was done or if memory was successfully resized. False to indicate that the allocator
+     *         could not provide enough memory to move the existing objects to a smaller allocation.
+     */
+    bool shrink_to_fit() noexcept(
+        noexcept(
+            allocator_type().deallocate(nullptr, 0)
+        )
+        &&
+        noexcept(
+            allocator_type().allocate(0)
+        )
+        &&
+        noexcept(
+            move_and_free<T>(nullptr, nullptr, 0, 0, std::declval<Allocator&>())
+        )
+    )
+    {
+        if (size_ == capacity_)
+        {
+            // already shrunk
+            return true;
+        }
+
+        // Special case where we are shrinking to empty
+        if (size_ == 0)
+        {
+            alloc_.deallocate(data_, capacity_);
+            data_     = nullptr;
+            capacity_ = 0;
+            return true;
+        }
+
+        // Allocate only enough to store what we have.
+        T* minimized_data = nullptr;
+
+        try
+        {
+            minimized_data = alloc_.allocate(size_ * sizeof(T));
+        }
+        catch (const std::bad_alloc& e)
+        {
+            // we ignore the exception since all allocation failures are modeled using
+            // null by this class.
+            (void) e;
+        }
+
+        if (minimized_data == nullptr)
+        {
+            return false;
+        }
+        else
+        {
+            if (minimized_data != data_)
+            {
+                move_and_free(minimized_data, data_, size_, capacity_, alloc_);
+            }  // else the allocator was able to simply shrink the reserved area for the same memory pointer.
+            data_     = minimized_data;
+            capacity_ = size_;
+            return true;
+        }
+    }
+
+    // +----------------------------------------------------------------------+
+    // | MODIFIERS
+    // +----------------------------------------------------------------------+
+    /**
+     * Push a new element on to the back of the array and grow the array size by 1.
+     *
+     * @return A pointer to the stored value or nullptr if there was not enough capacity (use reserve to
+     *         grow the available capacity).
```

Comment:
Why can't we have `try_push_back() -> T*` that invokes `reserve()` automatically?

Reply:
Because this isn't a generalized container type and I'd never generate code that relied on this. I want code that always calls reserve once to ensure the subsequent logic will be successful.

Reply:
Sure but we need to be application-friendly. Say, if I want to push X battery cell voltages into my array, calling `try_push_back()` X times is easier than `reserve()` once and then `push_back_no_alloc()` (both require error checking anyway).